File: erlang/lifegame.erl

-module(lifegame).
-export([main/0]).

main() ->
  Height = 20,
  Width = 40,
  Field = init_field(Height, Width),
  lifegame(Field, Height, Width).

init_field(Height, Width) ->
  List = [{{Y,X}, rand:uniform(2) - 1} || Y<-lists:seq(0, Height-1), X<-lists:seq(0, Width-1)],
  orddict:from_list(List).

lifegame(Field, Height, Width) ->
  clear_screen(),
  dump_field(Field, Height, Width),
  timer:sleep(100), % 100[ms]
  NewField = evolve(Field, Height, Width),
  lifegame(NewField, Height, Width).

evolve(Field, Height, Width) ->
  List = [{{Y,X}, dead_or_alive(Field, Height, Width, {Y,X})} || Y<-lists:seq(0, Height-1), X<-lists:seq(0, Width-1)],
  NewField = orddict:from_list(List),
  NewField.

dead_or_alive(Field, Height, Width, {Y,X}) ->
  AliveNeighbours = count_alive_neighbours(Field, Height, Width, {Y,X}),
  if
    AliveNeighbours =< 1 -> 0;
    AliveNeighbours == 2 -> orddict:fetch({Y,X}, Field);
    AliveNeighbours == 3 -> 1;
    AliveNeighbours >= 4 -> 0
  end.

count_alive_neighbours(Field, Height, Width, {Y,X}) ->
  Neighbours = [ orddict:fetch({mod(Y + Yi, Height), mod(X + Xi, Width)}, Field)
                || Yi<-[-1,0,1], Xi<-[-1,0,1], not ((Yi == 0) and (Xi == 0))],
  lists:sum(Neighbours).

mod(A, B) when A > 0 -> A rem B;
mod(A, B) when A < 0 -> mod(A + B, B);
mod(0, _) -> 0.

dump_field(Field, _, Width) ->
  Lines = chunks_of(Width, lists:map(fun to_char/1, orddict:to_list(Field))),
  io:format("~s~n", [ string:join(Lines, "\n") ]).

to_char({{_,_},Ch}) -> case Ch of 0 -> " "; 1 -> "o" end.

chunks_of(N, List) ->
  lists:foldr(fun(E, []) -> [[E]];
                 (E, [H|RAcc]) when length(H) < N -> [[E|H]|RAcc];
                 (E, [H|RAcc]) -> [[E],H|RAcc]
              end, [], List).

clear_screen() ->
  io:format("\033[;H\033[2J").

File: python/lifegame.py

import random
import time

def main():
    field = init_field(width=40, height=20)
    while True:
        clear_screen()
        dump_field(field)
        time.sleep(0.1) # 100[ms]
        field = evolve(field)

def init_field(height, width):
    return [ [ random.randint(0, 1) for x in range(width) ] for y in range(height) ]

def evolve(field):
    height, width = len(field), len(field[0])
    new_field = [ [ dead_or_alive(field, y, x) for x in range(width) ] for y in range(height) ]
    return new_field

def dead_or_alive(field, y, x):
    alive_neighbours = count_alive_neighbours(field, y, x)
    if alive_neighbours <= 1: return 0
    if alive_neighbours == 2: return field[y][x]
    if alive_neighbours == 3: return 1
    if alive_neighbours >= 4: return 0

def count_alive_neighbours(field, y, x):
    height, width = len(field), len(field[0])
    neighbours = [ field[(y + yi) % height][(x + xi) % width]
                   for yi in [-1,0,1] for xi in [-1,0,1] if not (yi == 0 and xi == 0) ]
    return sum(neighbours)

def dump_field(field):
    height, width = len(field), len(field[0])
    char_field = [ [ ' ' if field[y][x] == 0 else 'o' for x in range(width) ] for y in range(height) ]
    lines = [ ''.join(char_field[y]) for y in range(height) ]
    print("\n".join(lines))

def clear_screen():
    print("\033[;H\033[2J")

if __name__ == '__main__':
    main()
